monai/networks/trt_wrapper.py

```diff
@@ -0,0 +1,454 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import os
+import pickle
+import threading
+from collections import OrderedDict
+
+import torch
+
+from monai.apps.utils import get_logger
+from monai.networks.utils import add_casts_around_norms, convert_to_onnx
+from monai.utils.module import optional_import
+
+P, P_imported = optional_import("polygraphy")
+if P_imported:
+    from polygraphy.backend.common import bytes_from_path
+    from polygraphy.backend.onnxrt import OnnxrtRunner, session_from_onnx
+    from polygraphy.backend.trt import (
+        CreateConfig,
+        ModifyNetworkOutputs,
+        Profile,
+        engine_from_bytes,
+        engine_from_network,
+        network_from_onnx_path,
+        save_engine,
+    )
+
+trt, trt_imported = optional_import("tensorrt")
+cudart, _ = optional_import("cuda.cudart")
+
+LOGGER = get_logger("run_cmd")
+
+lock_sm = threading.Lock()
+
+
+# Map of TRT dtype -> Torch dtype
+def trt_to_torch_dtype_dict():
+    return {
+        trt.int32: torch.int32,
+        trt.float32: torch.float32,
+        trt.float16: torch.float16,
+        trt.bfloat16: torch.float16,
+        trt.int64: torch.int64,
+        trt.int8: torch.int8,
+        trt.bool: torch.bool,
+    }
+
+
+def get_dynamic_axes(profiles):
+    """
+    Given [[min,opt,max],...] list of profile dimensions,
+    this method calculates dynamic_axes to use in onnx.export()
+    """
+    dynamic_axes: dict[str, list[int]] = {}
+    if not profiles:
+        return dynamic_axes
+    for profile in profiles:
+        for key in profile:
+            axes = []
+            vals = profile[key]
+            for i in range(len(vals[0])):
+                if vals[0][i] != vals[2][i]:
+                    axes.append(i)
+            if len(axes) > 0:
+                dynamic_axes[key] = axes
+    return dynamic_axes
+
+
+def cuassert(cuda_ret):
+    """
+    Error reporting method for CUDA calls
+    """
+    err = cuda_ret[0]
+    if err != 0:
+        raise RuntimeError(f"CUDA ERROR: {err}")
+    if len(cuda_ret) > 1:
+        return cuda_ret[1]
+    return None
+
+
+class ShapeError(Exception):
+    """
+    Exception class to report errors from setting TRT plan input shapes
+    """
+
+    pass
+
+
+class Engine:
```

Comment:
Can we change this class name to `TRTEngine`? MONAI already has an `Engine` class for PyTorch ignite workflow:
https://github.com/Project-MONAI/MONAI/tree/dev/monai/engines

Thanks.

Reply:
Sure, that's a good idea.

Reply:
Done.